File: internal/protocol/canonical.go
package protocol

import (
	"bytes"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
)

// Canonicalize 把任意 JSON 可序列化值编码成协议 v1 规范化 JSON 字节。
// 规则见 docs/superpowers/plans/2026-09-25-base-p0-plan.md「接口契约补充」第 6 条。
func Canonicalize(v any) ([]byte, error) {
	raw, err := json.Marshal(v)
	if err != nil {
		return nil, fmt.Errorf("canonical json: marshal: %w", err)
	}
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	var x any
	if err := dec.Decode(&x); err != nil {
		return nil, fmt.Errorf("canonical json: decode: %w", err)
	}
	var buf bytes.Buffer
	if err := writeCanon(&buf, x); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

func writeCanon(buf *bytes.Buffer, v any) error {
	switch t := v.(type) {
	case nil:
		buf.WriteString("null")
	case bool:
		if t {
			buf.WriteString("true")
		} else {
			buf.WriteString("false")
		}
	case string:
		writeCanonString(buf, t)
	case json.Number:
		s := t.String()
		if strings.ContainsAny(s, ".eE") {
			return fmt.Errorf("canonical json: only integers allowed, got %q", s)
		}
		if s == "-0" {
			s = "0"
		}
		buf.WriteString(s)
	case []any:
		buf.WriteByte('[')
		for i, e := range t {
			if i > 0 {
				buf.WriteByte(',')
			}
			if err := writeCanon(buf, e); err != nil {
				return err
			}
		}
		buf.WriteByte(']')
	case map[string]any:
		keys := make([]string, 0, len(t))
		for k := range t {
			if !isASCII(k) {
				return fmt.Errorf("canonical json: non-ascii object key %q", k)
			}
			keys = append(keys, k)
		}
		sort.Strings(keys)
		buf.WriteByte('{')
		for i, k := range keys {
			if i > 0 {
				buf.WriteByte(',')
			}
			writeCanonString(buf, k)
			buf.WriteByte(':')
			if err := writeCanon(buf, t[k]); err != nil {
				return err
			}
		}
		buf.WriteByte('}')
	default:
		return fmt.Errorf("canonical json: unsupported type %T", v)
	}
	return nil
}

func isASCII(s string) bool {
	for i := 0; i < len(s); i++ {
		if s[i] >= 0x80 {
			return false
		}
	}
	return true
}

func writeCanonString(buf *bytes.Buffer, s string) {
	buf.WriteByte('"')
	for _, r := range s {
		switch r {
		case '"':
			buf.WriteString(`\"`)
		case '\\':
			buf.WriteString(`\\`)
		case '\b':
			buf.WriteString(`\b`)
		case '\f':
			buf.WriteString(`\f`)
		case '\n':
			buf.WriteString(`\n`)
		case '\r':
			buf.WriteString(`\r`)
		case '\t':
			buf.WriteString(`\t`)
		default:
			if r < 0x20 {
				fmt.Fprintf(buf, `\u%04x`, r)
			} else {
				buf.WriteRune(r)
			}
		}
	}
	buf.WriteByte('"')
}

File: packages/protocol-ts/src/canonical.ts
export type Json = null | boolean | number | string | Json[] | { [k: string]: Json };

/** 规范化 JSON v1（契约第 6 条），必须与 Go 侧 protocol.Canonicalize 字节一致。 */
export function canonicalize(value: Json): string {
  const out: string[] = [];
  writeCanon(value, out);
  return out.join("");
}

function writeCanon(v: Json, out: string[]): void {
  if (v === null) {
    out.push("null");
    return;
  }
  switch (typeof v) {
    case "boolean":
      out.push(v ? "true" : "false");
      return;
    case "string":
      writeCanonString(v, out);
      return;
    case "number": {
      if (!Number.isFinite(v) || !Number.isInteger(v)) {
        throw new Error(`canonical json: only integers allowed, got ${String(v)}`);
      }
      // 与 Go 一致：Go 的 json.Marshal(1e21) 产出 "1e+21" 会被拒，这里同样拒字符串化的指数形态
      const s = Object.is(v, -0) ? "0" : String(v);
      if (/[.eE]/.test(s)) {
        throw new Error(`canonical json: only integers allowed, got ${s}`);
      }
      out.push(s);
      return;
    }
    case "object": {
      if (Array.isArray(v)) {
        out.push("[");
        for (let i = 0; i < v.length; i++) {
          if (i > 0) out.push(",");
          writeCanon(v[i], out);
        }
        out.push("]");
        return;
      }
      const obj = v as { [k: string]: Json };
      const keys = Object.keys(obj);
      for (const k of keys) {
        if (!isASCII(k)) {
          throw new Error(`canonical json: non-ascii object key ${JSON.stringify(k)}`);
        }
      }
      keys.sort(); // 键全为 ASCII，码元序即 UTF-8 字节序
      out.push("{");
      for (let i = 0; i < keys.length; i++) {
        if (i > 0) out.push(",");
        writeCanonString(keys[i], out);
        out.push(":");
        writeCanon(obj[keys[i]], out);
      }
      out.push("}");
      return;
    }
    default:
      throw new Error(`canonical json: unsupported type ${typeof v}`);
  }
}

export function isASCII(s: string): boolean {
  for (let i = 0; i < s.length; i++) {
    if (s.charCodeAt(i) > 0x7f) return false;
  }
  return true;
}

function writeCanonString(s: string, out: string[]): void {
  out.push('"');
  for (let i = 0; i < s.length; i++) {
    const c = s.charCodeAt(i);
    switch (c) {
      case 0x22:
        out.push('\\"');
        break;
      case 0x5c:
        out.push("\\\\");
        break;
      case 0x08:
        out.push("\\b");
        break;
      case 0x0c:
        out.push("\\f");
        break;
      case 0x0a:
        out.push("\\n");
        break;
      case 0x0d:
        out.push("\\r");
        break;
      case 0x09:
        out.push("\\t");
        break;
      default:
        if (c >= 0xd800 && c <= 0xdbff) {
          const d = i + 1 < s.length ? s.charCodeAt(i + 1) : 0;
          if (d >= 0xdc00 && d <= 0xdfff) {
            out.push(s[i], s[i + 1]);
            i++;
            break;
          }
          out.push("\uFFFD"); // 与 Go 的 range over string 语义一致（孤立代理 → U+FFFD）
          break;
        }
        if (c >= 0xdc00 && c <= 0xdfff) {
          out.push("\uFFFD");
          break;
        }
        if (c < 0x20) {
          out.push("\\u" + c.toString(16).padStart(4, "0"));
          break;
        }
        out.push(s[i]);
    }
  }
  out.push('"');
}